### ISSUE TITLE
Add support for wildcard in the upper bound to improve Java interop

### ISSUE THREAD

**Issue by pusolito** [ceylon/ceylon-spec#1386]
Ceylon doesn't support wildcard in the upper bound which prevents interacts with Java like the following:

```java
// Java class that operates on java.lang.Comparable

public class A<T extends Comparable> {}
```
```ceylon
// Ceylon code to work with above Java type

import java.lang {Comparable}
import java.time {LocalDateTime}

A<T> bar<T>(A<T> t) given T satisfies Comparable<T> => t;

void test() {
    bar<LocalDateTime>(A<LocalDateTime>()); //<============= Error [Type parameter T of declaration
    bar has argument LocalDateTime not assignable to upper bound Comparable<Object> of T]
}
```

From @gavinking:
>
>Whoah: 
>
>```java
>public interface ChronoLocalDateTime<D extends ChronoLocalDate>  extends Temporal, >TemporalAdjuster, Comparable<ChronoLocalDateTime<?>>
>```
>
> OMG! 
>
> So it looks like what you're really wanting to write is this
>
>```ceylon
>A<T> bar<T>(A<T> t) given T satisfies Comparable<in T> => t; 
>```
>
>i.e. the wildcard in the supertype of ChronoLocalDateTime is obligating you to need a wildcard in the >upper bound, which is something Ceylon simply doesn't support, though it seems Java *does*.  (I never >noticed that before.)

>Open a feature request against ceylon-spec please, so that we can investigate if we can/should allow wildcards in upper bound type constraints.

**Comment by gavinking** [comment]
@RossTate thoughts? Are wildcards in upper bounds OK?

**Comment by RossTate** [comment]
Oops, I just saw this. As I mentioned in the thread, that's not necessary to solve the problem. You probably just have a glitch in importing the wildcard in the inheritance clause.

Here are the reductions that prove that the `LocalDateTime` case should be rejected and the `ChronoLocalDateTime<LocalTime>` cause should be accepted.

`LocalDateTime` case

1. `LocalDateTime <: Comparable<LocalDateTime>`
2. `Comparable<ChronoLocalDateTime<?>> <: Comparable<LocalDateTime>` (by inheritance)
3. `LocalDateTime <: ChronoLocalDateTime<?>` (by contravariance)
4. False - no such inheritance clause

`ChronoLocalDateTime<LocalTime>` case

1. `ChronoLocalDateTime<LocalTime> <: Comparable<ChronoLocalDateTime<LocalTime>>`
2. `Comparable<ChronoLocalDateTime<?>> <: Comparable<ChronoLocalDateTime<LocalTime>>` (by inheritance)
3. `ChronoLocalDateTime<LocalTime> <: ChronoLocalDateTime<?>` (by contravariance)
4. True by definition of `?`

**Comment by pusolito** [comment]
@RossTate,

To clarify, this following code does not compile with Ceylon IDE 1.1.1.v20150721-1337-Final.  The imports from Java are correct and verified.

Can you provide more specifics on how to get this working in the ChronoLocalDateTime<LocalTime> at least?  Or are you saying this doesn't work b/c of a separate bug within the import handling?

```ceylon
import java.lang {Comparable,Double}
import java.time {LocalDateTime, LocalDate}
import java.time.chrono {ChronoLocalDateTime}


A<T> bar<T>(A<T> t) given T satisfies Comparable<T> => t;

void test() {
    bar<Double>(A<Double>());
    bar<LocalDateTime>(A<LocalDateTime>());
    bar<ChronoLocalDateTime<LocalDate>>(A<ChronoLocalDateTime<LocalDate>>());
}
```

![image](https://cloud.githubusercontent.com/assets/9815928/8863221/8ffc94d0-314b-11e5-9554-6489982efafd.png)

**Comment by RossTate** [comment]
> Or are you saying this doesn't work b/c of a separate bug within the import handling?

Correct. And the feature proposed hear would not solve any of the problems. So I think this would be better classified as an issue for the compiler rather than for the spec. However, I'd first like @gavinking to chime in to confirm that my reasoning is sound.

Also, the `LocalDateTime` case will *never* work because it would be unsound to accept that. According to its definition, `LocalDateTime` is a subtype of `Comparable<ChronoLocalDateTime<?>>`, but that is not a subtype of `Comparable<LocalDateTime>` because `Comparable` is *contra*variant, not *co*variant (i.e. `in` not `out`).

**Comment by FroMage** [comment]
Actually, `java.lang.Comparable` is invariant since Java does not have declaration-site variance information.

**Comment by gavinking** [comment]
@RossTate 

> 1. `ChronoLocalDateTime<LocalTime> <: Comparable<ChronoLocalDateTime<LocalTime>>`

Where do you get that from. The declaration is:

```java
    public interface ChronoLocalDateTime<D extends ChronoLocalDate>
            extends Temporal, TemporalAdjuster, Comparable<ChronoLocalDateTime<?>> { ... }
```

i.e. `ChronoLocalDateTime<LocalTime>` is a subtype of `Comparable<ChronoLocalDateTime<?>>` but the typechecker is complaining that it's not a subtype of `Comparable<ChronoLocalDateTime<LocalTime>>`

**Comment by gavinking** [comment]
@RossTate 

> 2. `Comparable<ChronoLocalDateTime<?>> <: Comparable<ChronoLocalDateTime<LocalTime>>` (by inheritance)

I'm also not seeing this one. `Comparable` is an invariant Java type, and `ChronoLocalDateTime<?>` isn't the same type as `ChronoLocalDateTime<LocalTime>`. So I don't think that step is correct at all. It _would_ be correct for Ceylon's contravariant `Comparable`, but due to variance, not by inheritance.

> 3. `ChronoLocalDateTime<LocalTime> <: ChronoLocalDateTime<?>` (by contravariance)

This is sound, but not quite for the stated reason: we treat `ChronoLocalDateTime<?>` as `ChronoLocalDateTime<out Anything>` in the model loader. So actually it's true by _covariance_ ;-)

**Comment by RossTate** [comment]
You're misreading my proofs because I wrote them backwords. Read them as "in order to prove step 1, we must prove step 2, which then requires step 3, which then holds or doesn't hole due to step 4". The parentheticals indicate why the prior step reduces to that step.

Also, since it sounds like you're not inferring that Java's `Comparable` is contravariant, the signature of `bar` should be the following:
```A<T> bar<T>(A<T> t) given T satisfies Comparable<in T> => t;```

Try that. If that works, then the problem was that @pusolito gave `bar` an unnecessarilly restrictive signature. If that doesn't work, then there's a glitch in the type importer or type checker.

**Comment by gavinking** [comment]
> the signature of `bar` should be the following:
>    <T> bar<T>(A<T> t) given T satisfies Comparable<in T> => t;

Right, _precisely_! That's exactly what I wrote in the issue description!

The problem is that right now we never allow a wildcard instantiation of a supertype, not even for type parameters (i.e. upper bounds). So I can't write that.

**Comment by RossTate** [comment]
Ohh, sorry, I didn't realize you were calling `Comparable<in T>` in Ceylon a wildcard instantiation. (It's technically just a use-site annotation.) So, yes, you should allow use-site annotations in bounds. My proofs already handle that.

**Comment by gavinking** [comment]
@RossTate cool, thanks.
